services/queue/src/utils.js

```diff
@@ -0,0 +1,133 @@
+const assert = require('assert').strict;
+const { fromNow } = require('taskcluster-client');
+const { paginateResults } = require('taskcluster-lib-api');
+const debug = require('debug')('utils');
+
+const artifactUtils = {
+  // Create a single instance, or undefined, from a set of rows containing zero
+  // or one elements.
+  fromDbRows(rows) {
+    if (rows.length === 1) {
+      return exports.artifactUtils.fromDb(rows[0]);
+    }
+  },
+  // Create a single instance from a DB row
+  fromDb(row) {
+    return {
+      taskId: row.task_id,
+      runId: row.run_id,
+      name: row.name,
+      storageType: row.storage_type,
+      contentType: row.content_type,
+      details: row.details,
+      present: row.present,
+      expires: row.expires,
+      etag: row.etag,
+    };
+  },
+  // Create a serializable representation of this namespace suitable for response
+  // from an API method.
+  serialize(artifact) {
+    return {
+      storageType: artifact.storageType,
+      name: artifact.name,
+      expires: artifact.expires.toJSON(),
+      contentType: artifact.contentType,
+    };
+  },
+  //  Get all expired artifacts using a handler function.
+  // This is to avoid loading all epxired queue artifacts in memory.
+  async getExpiredArtifacts(db, expires, handler) {
+    assert(!handler || handler instanceof Function,
+      'If options.handler is given it must be a function');
+
+    const fetchResults = async (continuation) => {
+      const query = continuation ? { continuationToken: continuation } : {};
+      const {continuationToken, rows} = await paginateResults({
+        query,
+        fetch: (size, offset) => db.fns.get_queue_artifacts(
+          null,
+          null,
+          expires,
+          size,
+          offset,
+        ),
+      });
+
+      const entries = rows.map(exports.artifactUtils.fromDb);
+
+      return { rows: entries, continuationToken: continuationToken };
+    };
+
+    // Fetch results
+    let results = await fetchResults({});
+
+    const handleResults = async (res) => {
+      await Promise.all(res.rows.map((item) => handler.call(item, item)));
+
+      if (res.continuationToken) {
+        return await handleResults(await fetchResults(res.continuationToken));
+      }
+    };
+
+    results = await handleResults(results);
+
+    return results;
+  },
+  /**
+   * Remove underlying artifact and entity representing it.
+   *
+   * This method will remove both the artifact in the db and underlying artifact.
+   * But the artifact in the db will not be deleted if there is an error
+   * deleting the underlying artifact.
+   */
+  async expire({ db, publicBucket, privateBucket, ignoreError, monitor, expires }) {
+    let count = 0;
+    await exports.artifactUtils.getExpiredArtifacts(db, expires, async (artifact) => {
```

Comment:
This seems to be the only use of `getExpiredArtifacts`, so perhaps that utility could be combined here, or incorporated as a nested function?

I think this is going to miss expiring artifacts:

* call get_queue_artifacts with a page size of 1000
* call S3 for each of those 1000
* call delete_artifact for each of those 1000
* call get_queue_artifacts with a page size of 1000, offset 1000

..at this point, we've missed 1000 expired artifacts since the offset skips 1000 non-deleted artifacts.

This actually isn't the end of the world -- those artifacts would hopefully just get picked up in the next run of the expiration job -- but might be confusing when we wonder why there's a small tail of expired artifacts in the DB.

I suspect the fix would be to decrement the offset by 1 for every call to `delete_queue_artifact`.  In the success case, that will end up calling `get_queue_artifacts` with offset = 0 every time, until it returns no results, but in the failure case it will not re-try deleting S3 objects.

Separately, it seems like we should be taking a closer look at these errors instead of just ignoring them.  I'll file a followup for that.

Reply:
(#3183)